x/gamm/pool-models/stableswap/amm_test.go

```diff
@@ -12,6 +13,253 @@ import (
 	"github.com/osmosis-labs/osmosis/v12/x/gamm/pool-models/internal/test_helpers"
 )
 
+// multiAssetCFMMTestCase defines a testcase for TestCFMMInvariantTwoAssets
+// and TestCFMMInvariantTwoAssetsDirect
+type multiAssetCFMMTestCase struct {
+	xReserve    osmomath.BigDec
+	yReserve    osmomath.BigDec
+	uReserve    osmomath.BigDec
+	wSumSquares osmomath.BigDec
+	yIn         osmomath.BigDec
+	expectPanic bool
+}
+
+var multiAssetCFMMTestCases = map[string]multiAssetCFMMTestCase{
+	// sanity checks
+	"even 3-asset small pool, small input": {
+		xReserve: osmomath.NewBigDec(100),
+		yReserve: osmomath.NewBigDec(100),
+		// represents a 3-asset pool with 100 in each reserve
+		uReserve:    osmomath.NewBigDec(100),
+		wSumSquares: osmomath.NewBigDec(10000),
+		yIn:         osmomath.NewBigDec(1),
+		expectPanic: false,
+	},
+	"even 3-asset medium pool, small input": {
+		xReserve: osmomath.NewBigDec(100000),
+		yReserve: osmomath.NewBigDec(100000),
+		// represents a 3-asset pool with 100,000 in each reserve
+		uReserve:    osmomath.NewBigDec(100000),
+		wSumSquares: osmomath.NewBigDec(10000000000),
+		yIn:         osmomath.NewBigDec(100),
+		expectPanic: false,
+	},
+	"even 4-asset small pool, small input": {
+		xReserve: osmomath.NewBigDec(100),
+		yReserve: osmomath.NewBigDec(100),
+		// represents a 4-asset pool with 100 in each reserve
+		uReserve:    osmomath.NewBigDec(10000),
+		wSumSquares: osmomath.NewBigDec(20000),
+		yIn:         osmomath.NewBigDec(1),
+		expectPanic: false,
+	},
+	"even 4-asset medium pool, small input": {
+		xReserve: osmomath.NewBigDec(100000),
+		yReserve: osmomath.NewBigDec(100000),
+		// represents a 4-asset pool with 100,000 in each reserve
+		uReserve:    osmomath.NewBigDec(10000000000),
+		wSumSquares: osmomath.NewBigDec(20000000000),
+		yIn:         osmomath.NewBigDec(1),
+		expectPanic: false,
+	},
+	/* TODO: increase BigDec precision (36 -> 72) to be able to accommodate this
```

Comment:
random question: how much is this going to affect performance? I'm assuming this will make some math operations to converge slower. Do we have benchmarks / should we be comparing trade-offs?

Reply:
I'm not sure about performance impact, but the max number of assets per pool we want to support is still an open question that will determine the final precision (72 isn't enough for large 5+ asset pools)

Once we finalize precision, we can expand `amm_bench_test.go` tests to see what the performance effect is, but it is upper bounded at `maxIterations` so it likely won't be too bad